### PR TITLE
feat: Support Cargo alternate registries in aas renovate

### PR DESCRIPTION
**Pull request by rbtcollins**
## Changes

WIP

## Documentation (please check one with an [x])

- [X] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: docs/usage/rust.md [hunk 1]
```diff
@@ -19,11 +19,18 @@ Renovate supports upgrading dependencies in `Cargo.toml` files and their accompa
...
 Third-party crate registries usually provide instructions to achieve this.
 
 Both of these are currently only possible when running Renovate self-hosted.
+
+## Private sparse index registries
+
+If dependencies are available via a sparse index, then a `registryAlias` can be
+used to have Renovate accesss the dependencies via the sparse protocol. Use
+`HostRules` to provide authentication details if needed. Crates-io index access
+will still be via Renovates shared clone of the public git index.
```
Comment (HonkingGoose) [inline docs/usage/rust.md:36]:
```suggestion
If dependencies are available via a "sparse index", you may use `registryAliases` to let Renovate accesses the dependencies via the `sparse` protocol.
If needed, set the authentication details with the `hostRules` config option.
`crates.io` index access will still be via Renovates shared clone of the public Git index.
```

Rewrite:
  - Use `you` to draw the reader into the text
  - Put "sparse index" between quotes, to highlight the term
  - Use simple words and phrases
  - One sentence per line
  - Use `crates.io` to match line 26 of this file
  
Grammar/typos:

- The `hostRules` config option name starts with a normal case latter
- The config option is `registryAliases` not `registryAlias`
- Git is a proper noun, so must be capitalized

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/rust.md [hunk 2]
```diff
@@ -30,7 +30,46 @@ Both of these are currently only possible when running Renovate self-hosted.
...
 
-If dependencies are available via a sparse index, then a `registryAlias` can be
-used to have Renovate accesss the dependencies via the sparse protocol. Use
-`HostRules` to provide authentication details if needed. Crates-io index access
-will still be via Renovates shared clone of the public git index.
+If dependencies are available via a "sparse index", you may use
+`registryAliases` to let Renovate accesses the dependencies via the `sparse`
+protocol. If needed, set the authentication details with the `hostRules` config
+option. `crates.io` index access will still be via Renovates shared clone of the
+public Git index.
```
Comment (HonkingGoose) [inline docs/usage/rust.md:37]:
```suggestion
If dependencies are available via a "sparse index", you may use `registryAliases` to let Renovate accesses the dependencies via the `sparse` protocol.
If needed, set the authentication details with the `hostRules` config option.
 `crates.io` index access will still be via Renovates shared clone of the public Git index.
```

File: docs/usage/rust.md [hunk 3]
```diff
@@ -30,7 +30,46 @@ Both of these are currently only possible when running Renovate self-hosted.
...
+
+For instance with a username and password:
+
+```
```
Comment (HonkingGoose) [inline docs/usage/rust.md:41]:
```suggestion
```json
```

File: docs/usage/rust.md [hunk 4]
```diff
@@ -30,7 +30,46 @@ Both of these are currently only possible when running Renovate self-hosted.
...
+      "matchHost": "https://INSTANCE.jfrog.io",
+      "authType": "Token-Only",
+      "token": "Basic username-and-password-base64-encoded"
+    }
+  ]
+```
+
+With a bearer token:
+
+```
```
Comment (HonkingGoose) [inline docs/usage/rust.md:60]:
```suggestion
```json
```

File: package.json [hunk 5]
```diff
@@ -166,6 +166,7 @@
     "auth-header": "1.0.0",
     "aws4": "1.11.0",
     "azure-devops-node-api": "11.2.0",
+    "bsip": "0.1.9",
```
Comment (HonkingGoose) [inline package.json:169]: I don't see any code changes in this PR to actually load this dependency? I was expecting to see some changed TypeScript files.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Docs need some more changes. Are you sure about the `bsip` dependency?

File: package.json (hunk 5 above)
Comment (rbtcollins) [reply package.json:169]: Hi, the PR is in draft state - we're discussing the design in https://github.com/renovatebot/renovate/discussions/18872 - and as I get things I'm sure of I'm committing them and pushing. 

So there's no need to review it per se until I remove the Draft marker - though I do absolutely appreciate the input.

The bsip dependency will be used to determine the correct file name in the cargo registry dir for a given registry.

File: docs/usage/rust.md [hunk 6]
```diff
@@ -19,11 +19,57 @@ Renovate supports upgrading dependencies in `Cargo.toml` files and their accompa
...
 
 Both of these are currently only possible when running Renovate self-hosted.
+
+## Private sparse index registries
+
+If dependencies are available via a "sparse index", you may use
+`registryAliases` to let Renovate accesses the dependencies via the `sparse`
+protocol. If needed, set the authentication details with the `hostRules` config
+option. `crates.io` index access will still be via Renovates shared clone of
+the public Git index.
```
Comment (HonkingGoose) [inline docs/usage/rust.md:37]:
```suggestion
If dependencies are available via a "sparse index", you may use `registryAliases` to let Renovate accesses the dependencies via the `sparse` protocol.
If needed, set the authentication details with the `hostRules` config option.
`crates.io` index access will still be via Renovate's shared clone of the public Git index.
```

We use the "one sentence per line" pattern in our docs. 😉

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
Closing due to inactivity. This topic is welcome to be reopened in future!

**Pull request closed without merging**